lightning/src/util/ser.rs

```diff
@@ -34,65 +34,71 @@ use util::byte_utils::{be64_to_array, be48_to_array, be32_to_array, be16_to_arra
 /// serialization buffer size
 pub const MAX_BUF_SIZE: usize = 64 * 1024;
 
-/// A trait that is similar to std::io::Write but has one extra function which can be used to size
-/// buffers being written into.
-/// An impl is provided for any type that also impls std::io::Write which simply ignores size
-/// hints.
+/// A struct that implements a serialization buffer.  A size hint can be given and the length
+/// of the result can be retrieved.
 ///
-/// (C-not exported) as we only export serialization to/from byte arrays instead
-pub trait Writer {
+/// std::io::Write is implemented for serializing Bitcoin objects with impl_consensus_ser! below.
+pub struct Writer(pub Vec<u8>);
+
+impl Writer {
+	/// Construct an empty Writer
+	pub fn new() -> Self {
+		Writer(Vec::new())
+	}
+
 	/// Writes the given buf out. See std::io::Write::write_all for more
-	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error>;
+	pub fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.extend_from_slice(buf);
+		Ok(())
+	}
+
 	/// Hints that data of the given size is about the be written. This may not always be called
 	/// prior to data being written and may be safely ignored.
-	fn size_hint(&mut self, size: usize);
-}
-
-impl<W: Write> Writer for W {
-	#[inline]
-	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
-		<Self as ::std::io::Write>::write_all(self, buf)
+	pub fn size_hint(&mut self, size: usize) {
+		self.0.reserve(size)
+	}
+
+	/// Return the number of bytes written
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	// Write the data with a varint length prefix.
+	// Use length_adjust to add to the length for special cases, such as writing an Option.
+	pub(crate) fn write_all_with_length<T: Writeable>(&mut self, data: &T, length_adjust: usize) -> Result<(), ::std::io::Error> {
+		// Most common case is a single byte length prefix, so leave a space for that
+		self.0.push(0);
+		let start = self.len();
+		data.write(self).expect("No in-memory data may fail to serialize");
+		self.insert_length(start, length_adjust)
+	}
+
+	fn insert_length(&mut self, slice_start: usize, length_adjust: usize) -> Result<(), ::std::io::Error> {
+		let slice_end = self.len();
+		let length = slice_end - slice_start + length_adjust;
+		let big_size = BigSize(length as u64);
+		let len_len = big_size.expected_length();
```

Comment:
nit: what do you think about "big_size_len" ?

Reply:
Done